### ISSUE TITLE
deprecate `inject` and `Provider`

### ISSUE THREAD

**Issue by 8balloon** [mobxjs/mobx-react#410]
TL;DR: I would LOVE to break `inject` and `Provider` out into an independent library. 

Mobx has been lifechanging for me. Among other reasons, it's freed me from having to worry about hidden context and store:props mapping. I can import stores and reference them in `observer`s, and it's __very__ clear where state is coming from.

So I hope you can understand how confused I feel when I see that half of the docs and issues of this project are devoted to supporting `inject` and `Provider`. These utilities aren't necessary to use mobx-react -- in fact, I believe that they are harmful! They introduce the store:props mapping that makes other libraries so heinous, and they make dependencies and types way less apparent and easier to break. 

Is there some advantage to these utilities that I'm not getting? Because if it's just to avoid having to write a few `import` statements, it seems to come at a very high cost.

Example:

```js
// 1.
import counterStore from '../stores/counterStore'
const Counter = observer(() => {
    const { count } = counterStore
    return <div>{count}</div>
})

// 2.
import { inject } from 'mobx-react'
const Counter = inject('counterStore')(observer((props) => {
    const { count } = props.counterStore
    return <div>{count}</div>
}))
```

In example 2, you have to load the whole "idea" of `Provider`/`injector` into your head to understand where your state is coming from. Maybe it will save you a few lines, but is that worth it?

Just to add to my point, I've introduced two people to Mobx who were completely thrown off by `Provider` and `inject` until I explained that they are completely optional and unnecessary.

Maybe there's some sort of history behind the two that justifies keeping them as part of mobx-react?

Regardless, thank you so much for your work.

**Comment by urugator** [comment]
Using import is easy, but not clean design wise. It's the same as putting things into `globals` (with better namespace handling) or making everything `static`. Once you make one thing static, everything it depens on must be static as well. So you will end up with very rigid architecture, where everything is hardwired, imposible to compose and hard to test.   
However I agree that `Provider/@inject` should be in seperate package.

**Comment by Gvozd** [comment]
@8balloon There are tasks where contexts are more convenient than global store or regular props

e.g. in html editor you have some selected element at moment

Until you have one selection at the time:
1. You can store it globally, and import in every component. It's comfortable
2. You can store it in root Provider, and inject in every component . It's also comfortable
3. You can pass it to root comonent, and re-pass to props on every level. It's some painful(you need pass selection-prop to every child), but it works

But, when you have a more complex selection
e.g. when you selected some element, and want  edit his color, and color of **his parent**
Or you selected some complex element(CommentForm e.g), and want editable his sub-elements(textarea, button, ...)
1. You can not store selection globally, because you have more than one selection
2. You can store base selection in root Provider, and create additional Providers(with changed selection), where you want work with another elements(parents/childs/...). Imho, it's best variant
3. You can still use direct selection prop at every level, and change where you need.

**Comment by urugator** [comment]
@Gvozd You can avoid the need to pass props transitively through multiple levels by using render callbacks (actually you won't avoid this pattern if you want to make your components individually testable see #315 #366).

`@inject` shares some problems with `import`, but there are two differences:
1. The problems are not app wide, but isolated in specific subtree of components.
2. The dependency is not hardwired - you can still pass the depedency by prop, which makes the component itself (without the inject part) reusable and easier to test.

**Comment by 8balloon** [comment]
@urugator @Gvozd Thank you both for your thoughtful replies.

The points about dependency injection make a lot of sense to me. And it seems like it would work great for complex setups like @Gvozd described.

So far this all seems to support the case for breaking out `inject` into its own library... Doing so would make the delineation of use-cases very clear, and it would also  make onboarding easier, since users wouldn't have to figure out `inject` at the same time as the rest of Mobx.

Does @mweststrate have any input on this? Is there any reason not to break this feature out into its own library as part of Mobx 5?

**Comment by mweststrate** [comment]
@8balloon I think dependency injection is invaluable in complex project setups. Without it (when using global imports), it is not possible to do something as trivial as running unit tests in parallel. 

Beyond that you are right that Provider / inject in essence has nothing to do with MobX. In fact, the original blog post describing the pattern AFAIK inspired the development of libraries like react-broadcast, addressing the same issue.

And there is good news, the new context mechanism in React solves the root problem that caused Provider / inject to be added to this library: the incompabitility of Context in combination with `shouldComponentUpdate`. That means that as far as I am concerned Provider / inject could be removed in a next major. Not yet, because the new api is not ironed out completely (e.g. injecting multiple stores would be tricky). But I assume that at some point a library will emerge that follows the exact same pattern as Provider / inject in mobx-react.

**Comment by 8balloon** [comment]
@mweststrate Awesome :)

Please let me know if you'd like help with this.

**Comment by Gvozd** [comment]
@mweststrate 
>  problem that caused Provider / inject to be added to this library: the incompabitility of Context in combination with shouldComponentUpdate

I tried to solve this problem in #411 , and it looks like I did it
May you look my PR?

**Comment by mweststrate** [comment]
For now, figured to not deprecate the `Provider`/ `inject` api in the foreseeable future, because in quite some aspects that api is still more ergonomic then the React.createContext api, and forcing all consumers to upgrade would be quite painful

**Comment by 8balloon** [comment]
@mweststrate I've given this some more thought, especially with regards to writing testable code.

Testing frameworks like Jest allow mocking at the module level. Wouldn't this grant you the capacity for full-featured testing without having to access dependencies through a context/provider?

I feel like I'm in the minority on this one, but it's important to me, because I think that code becomes vastly less intelligible when module-level dependencies aren't explicitly declared. (For the record, I feel like the createContext API is a step backward.)

I think it could hugely benefit the community to move away from contexts/providers and towards module-level mocking... 

If you think I'm wrong on this, I would really appreciate an explanation :)

**Comment by urugator** [comment]
@8balloon You're missing the point ... nobody claims that testing is imposible, but more complicated. But that isn't the problem itself, just it's consequence.
As mentioned the problem is that everything is statically bound to module loader, so you can't control the "lifecycle" of objects and it's dependencies programatically.
It greatly reduces the reusability/composability of your code. It is a problem in general, which you may not care about from the perspective of the application logic itself, but in practice there are two consequences:
1) **You can't run your code in paralell**
Therefore can't run multiple tests in paralell on single VM.
Therefore you can't do SSR, because you can't serve multiple requests at the same time (module loader is shared by requests).
2) **You have to modify module loader logic to provide different dependencies or to re-set the state of the depedencies.**
Afaik this is what test libs do. They modifies the implementation of `require` function. 
Note that this solution won't work when you start using native `import` impl, without transpiling `import` statements into `require` statements. 
And because native import statements are not resolved dynamically like require, but statically, chances are that each individual test would have to start new VM instance, passing custom module loader hook as exec option in order to mock the module dependencies.

Also I would like to point out that context/providers (or any DI solution in general) is not an alternative to module imports, but an alternative to passing dependencies to *constructor functions*.

**Comment by 8balloon** [comment]
@urugator Thank you so much for your detailed explanation.

The solution you described with multiple VM instances is what I was personally looking for. The SSR use case you described explains why it wouldn't work as a universal solution. So I am satisfied :)

I'm sorry if I've been frustrating anyone with this. Hopefully this conversation thread helps others.

Thanks for the awesome library! I appreciate you guys.

**Comment by voold** [comment]
So is it better to use the React context api or the Provider/ inject from react/mobx? What are the pro/cons?

**Comment by melbourne2991** [comment]
Wouldn't it be trivial to build Provider/ Inject from the new context api?

@inject would look something like this:

```
const inject = (propNames) => (Component) => {
   return (<Context.Consumer>{(contextValue) => <Component 
  {...Object.keys(contextValue).map(propName => propNames.includes(propName))}/>} 
 </Context.Consumer>
}

}
```

If anything breaking it out would be nice because it's a handy helper function that I believe can be used for things that aren't stores/mobx related.

**Comment by urugator** [comment]
@melbourne2991 The provider would overwrite previously provided context value, you would have to wrap it into consumer and merge the value.
The inject should return component (and merge props with injected props), you're returning element...
Note that all consumers will be invoked whenever any Provider is invoked, this allows changing context dynamically, but could cause some perfomance degradation.
Why not to stick with one static context per value, what's the benefit of merging them into one and referencing them via strings? Note that you can implement helper function to merge multiple consumers like
```javascript
const ComponentWithStores = Consumer({
  store1: Store1Context,
  sotre2: Store2Context,
})(Component)
```
Obviously the same can be done for providers...

**Comment by zxiest** [comment]
Something to note is that on server-side rendering, importing the store (a singleton) can cause race conditions (ie: different "requests" hitting the same store), hence returning incorrect data.

**Comment by szagi3891** [comment]
I use the new context of the react. No problems, everything is well predicted.
